go/arrow/flight/flightsql/driver.go

```diff
@@ -0,0 +1,772 @@
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, d.timeout)
+		defer cancel()
+	}
+
+	client, err := NewClientCtx(ctx, d.addr, nil, nil, d.options...)
+	if err != nil {
+		return nil, err
+	}
+	d.client = client
+
+	return d, nil
+}
+
+// Configure the driver with the corresponding config
+func (d *Driver) Configure(config *DriverConfig) error {
+	// Set the driver properties
+	d.addr = config.Address
+	d.timeout = config.Timeout
+
+	// Create GRPC options necessary for the backend
+	var err error
+	switch config.Backend {
+	case "sqlite":
+		d.options, err = newSqliteBackend(config)
+		if err != nil {
+			return err
+		}
+	case "iox", "ioxs":
+		d.options, err = newIOxBackend(config)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("invalid backend %q", config.Backend)
+	}
+	d.options = append(d.options, grpc.WithBlock())
+
+	return nil
+}
+
+// Driver returns the underlying Driver of the Connector,
+// mainly to maintain compatibility with the Driver method
+// on sql.DB.
+func (d *Driver) Driver() driver.Driver {
+	return d
+}
+
+// Prepare returns a prepared statement, bound to this connection.
+func (d *Driver) Prepare(query string) (driver.Stmt, error) {
+	return d.PrepareContext(context.Background(), query)
+}
+
+// PrepareContext returns a prepared statement, bound to this connection.
+// context is for the preparation of the statement,
+// it must not store the context within the statement itself.
+func (d *Driver) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	if _, set := ctx.Deadline(); !set && d.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, d.timeout)
+		defer cancel()
+	}
+
+	var err error
+	var stmt *PreparedStatement
+	if d.txn != nil && d.txn.txn.IsValid() {
+		stmt, err = d.txn.Prepare(ctx, query)
+	} else {
+		stmt, err = d.client.Prepare(ctx, query)
+		d.txn = nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &Stmt{
+		stmt:    stmt,
+		client:  d.client,
+		timeout: d.timeout,
+	}, nil
+}
+
+// Close invalidates and potentially stops any current
+// prepared statements and transactions, marking this
+// connection as no longer in use.
+func (d *Driver) Close() error {
+	d.addr = ""
+	d.options = nil
+	if d.client == nil {
+		return nil
+	}
+
+	if d.txn != nil && d.txn.txn.IsValid() {
+		return ErrTransactionInProgress
+	}
+
+	// Drivers must ensure all network calls made by Close
+	// do not block indefinitely (e.g. apply a timeout).
+	err := d.client.Close()
+	d.client = nil
+
+	return err
+}
+
+// Begin starts and returns a new transaction.
+func (d *Driver) Begin() (driver.Tx, error) {
+	return d.BeginTx(context.Background(), sql.TxOptions{})
+}
+
+func (d *Driver) BeginTx(ctx context.Context, opts sql.TxOptions) (driver.Tx, error) {
+	tx, err := d.client.BeginTransaction(ctx)
+	if err != nil {
+		return nil, err
+	}
+	d.txn = tx
+
+	return &Tx{tx: tx, timeout: d.timeout}, nil
+}
```

Comment:
storing the transaction at the driver level means that this will break if there are multiple connections with their own transactions. You need a separate connection object and to store the transaction at the connection level instead, not the driver level.

Reply:
From what I read about the Golang code there is only one connection per `Driver` instance. Golang will then create a pool of drivers which are assigned to an transaction. There is no way to access the "connection" at the `BeginTx` call... What do you have in mind?

Reply:
> From what I read about the Golang code there is only one connection per Driver instance. Golang will then create a pool of drivers which are assigned to an transaction.

That's not quite the case. The `database/sql` package creates one driver instance and maintains a pool of connections. You're guaranteed that a given connection will only be used by one goroutine **at a time**, but you're not guaranteed it won't be used by multiple goroutines in general.

> There is no way to access the "connection" at the BeginTx call... What do you have in mind?

The `Open` function on the Driver should return a connection to the database, not the driver itself. That's where your issue is. `BeginTx` is a method on the `driver.Conn` interface, not `driver.Driver`. You need to fix the implementation to have an actual "connection" type. I highly recommend looking over https://github.com/apache/arrow-adbc/blob/main/go/adbc/sqldriver/driver.go for inspiration.